File: packages/ts/src/canonicalpath/sanitize.ts
import { createHash } from "node:crypto";
import { pathError } from "./errors.js";
import { isReservedDeviceBase } from "./internal.js";

export function sanitizeComponent(name: string, profile: "portable" | "win32" | "posix"): string {
  if (name === "") throw pathError("ERR_INVALID_COMPONENT", "component is empty");
  if (name.includes("\0")) throw pathError("ERR_NUL_BYTE", "component contains NUL");
  let value = name.replace(/[\\/:\t\n\r]+/g, "-").replace(/^[ ._-]+|[ ._-]+$/g, "");
  if (value === "") value = "component";
  if (profile === "win32") value = escapeReservedWin32Component(value);
  return value;
}

export function encodeComponent(name: string, profile: "portable" | "win32" | "posix"): string {
  return sanitizeComponent(name, profile);
}

export function encodeGitRef(raw: string): string {
  if (raw === "") throw pathError("ERR_INVALID_COMPONENT", "git ref is empty");
  if (raw.includes("\0")) throw pathError("ERR_NUL_BYTE", "git ref contains NUL");
  const slug = raw.replace(/[^A-Za-z0-9._-]+/g, "-").replace(/^[._-]+|[._-]+$/g, "") || "ref";
  const hash = createHash("sha256").update(raw, "utf8").digest("hex").slice(0, 12);
  return `${slug}--${hash}`;
}

function escapeReservedWin32Component(value: string): string {
  const dot = value.indexOf(".");
  const base = dot >= 0 ? value.slice(0, dot) : value;
  const suffix = dot >= 0 ? value.slice(dot) : "";
  if (isReservedDeviceBase(base.toUpperCase())) return `${base}-${suffix}`;
  return value;
}

File: packages/go/canonicalpath/sanitize.go
package canonicalpath

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

// SanitizeComponent converts a user string into a filesystem-friendly component.
func SanitizeComponent(name string, profile string) (string, error) {
	if name == "" {
		return "", newError(ErrInvalidComponent, "component is empty")
	}
	if strings.ContainsRune(name, '\x00') {
		return "", newError(ErrNULByte, "component contains NUL")
	}
	replacer := strings.NewReplacer("/", "-", "\\", "-", ":", "-", "\t", "-", "\n", "-", "\r", "-")
	value := strings.Trim(replacer.Replace(name), " ._-")
	if value == "" {
		value = "component"
	}
	if profile == "win32" {
		value = escapeReservedWin32Component(value)
	}
	return value, nil
}

// EncodeComponent encodes a user string as a component for the selected profile.
func EncodeComponent(name string, profile string) (string, error) {
	return SanitizeComponent(name, profile)
}

// EncodeGitRef encodes a Git ref into a collision-resistant directory name.
func EncodeGitRef(raw string) (string, error) {
	if raw == "" {
		return "", newError(ErrInvalidComponent, "git ref is empty")
	}
	if strings.ContainsRune(raw, '\x00') {
		return "", newError(ErrNULByte, "git ref contains NUL")
	}
	slug := slugGitRef(raw)
	hash := sha256.Sum256([]byte(raw))
	return slug + "--" + hex.EncodeToString(hash[:])[:12], nil
}

func slugGitRef(raw string) string {
	var builder strings.Builder
	lastDash := false
	for _, r := range raw {
		allowed := r >= 'a' && r <= 'z' || r >= 'A' && r <= 'Z' || r >= '0' && r <= '9' || r == '.' || r == '_' || r == '-'
		if allowed {
			builder.WriteRune(r)
			lastDash = false
			continue
		}
		if !lastDash {
			builder.WriteByte('-')
			lastDash = true
		}
	}
	slug := strings.Trim(builder.String(), "._-")
	if slug == "" {
		return "ref"
	}
	return slug
}

func escapeReservedWin32Component(value string) string {
	base := value
	suffix := ""
	if idx := strings.Index(base, "."); idx >= 0 {
		base = value[:idx]
		suffix = value[idx:]
	}
	if isReservedDeviceBase(strings.ToUpper(base)) {
		return base + "-" + suffix
	}
	return value
}
